File: CSIntroNeuralNetworkEdition2/HeatonResearchSpider/HTML/HTMLTag.cs
// The Heaton Research Spider for .Net 
// Copyright 2007 by Heaton Research, Inc.
// 
// From the book:
// 
// HTTP Recipes for C# Bots, ISBN: 0-9773206-7-7
// http://www.heatonresearch.com/articles/series/20/
// 
// This class is released under the:
// GNU Lesser General Public License (LGPL)
// http://www.gnu.org/copyleft/lesser.html
//
using System;
using System.Collections.Generic;
using System.Text;

namespace HeatonResearch.Spider.HTML
{
    
    /// <summary>
    /// HTMLTag: This class holds a single HTML tag. This class
    /// subclasses the AttributeList class. This allows the
    /// HTMLTag class to hold a collection of attributes, just as
    /// an actual HTML tag does.
    /// </summary>
    public class HTMLTag
    {
        private String name;
        private bool ending;

        /// <summary>
        /// The name of the tag.
        /// </summary>
        public String Name
        {
            get
            {
                return name;
            }
            set
            {
                name = value;
            }
        }

        /// <summary>
        /// Is this tag both a beginning and an
        /// ending tag.
        /// </summary>
        public Boolean Ending
        {
            get
            {
                return ending;
            }
            set
            {
                ending = value;
            }
        }

        /// <summary>
        /// The attributes of this tag.
        /// </summary>
        private Dictionary<String, String> attributes = new Dictionary<String, String>();

        /// <summary>
        /// Clear out this tag.
        /// </summary>
        public void Clear()
        {
            this.attributes.Clear();
            this.Name = "";
            this.Ending = false;
        }

        /// <summary>
        /// Access the individual attributes by name.
        /// </summary>
        public String this[string key]
        {
            get
            {
                if( attributes.ContainsKey(key.ToLower()) )
                    return this.attributes[key.ToLower()];
                else
                    return null;
            }
            set
            {
                this.attributes.Add(key.ToLower(), value);
            }
        }

        /// <summary>
        /// Convert this tag back into string form, with the
        /// beginning &lt; and ending &gt;.
        /// </summary>
        /// <returns>The attribute value that was found.</returns>
        public override String ToString()
        {
            StringBuilder buffer = new StringBuilder("<");
            buffer.Append(this.Name);

            foreach (String key in attributes.Keys)
            {
                String value = this.attributes[key];
                buffer.Append(' ');

                if (value == null)
                {
                    buffer.Append("\"");
                    buffer.Append(key);
                    buffer.Append("\"");
                }
                else
                {
                    buffer.Append(key);
                    buffer.Append("=\"");
                    buffer.Append(value);
                    buffer.Append("\"");
                }
            }

            if (this.Ending)
            {
                buffer.Append('/');
            }
            buffer.Append(">");
            return buffer.ToString();
        }

        /// <summary>
        /// Set the specified attribute.
        /// </summary>
        /// <param name="key">The attribute name.</param>
        /// <param name="value">The attribute value.</param>
        public void SetAttribute(String key, String value)
        {
            attributes.Remove(key.ToLower());
            attributes.Add(key.ToLower(), value);
        }
    }
}

File: JavaHTTPRecipesEdition1/com/heatonresearch/httprecipes/html/HTMLTag.java
package com.heatonresearch.httprecipes.html;

import java.util.*;

/**
 * The Heaton Research Spider Copyright 2007 by Heaton
 * Research, Inc.
 * 
 * HTTP Programming Recipes for Java ISBN: 0-9773206-6-9
 * http://www.heatonresearch.com/articles/series/16/
 * 
 * HTMLTag: This class holds a single HTML tag. This class
 * subclasses the AttributeList class. This allows the
 * HTMLTag class to hold a collection of attributes, just as
 * an actual HTML tag does.
 * 
 * This class is released under the:
 * GNU Lesser General Public License (LGPL)
 * http://www.gnu.org/copyleft/lesser.html
 * 
 * @author Jeff Heaton
 * @version 1.1
 */
public class HTMLTag
{
  /*
   * The attributes
   */
  private Map<String, String> attributes = new HashMap<String, String>();

  /**
   * The tag name.
   */
  private String name = "";

  /*
   * Is this both a beginning and ending tag.
   */
  private boolean ending;

  public void clear()
  {
    this.attributes.clear();
    this.name = "";
    this.ending = false;
  }

  /**
   * Get the value of the specified attribute.
   * 
   * @param name
   *          The name of an attribute.
   * @return The value of the specified attribute.
   */
  public String getAttributeValue(String name)
  {
    return this.attributes.get(name.toLowerCase());
  }

  /**
   * Get the tag name.
   */
  public String getName()
  {
    return this.name;
  }

  /**
   * @return the ending
   */
  public boolean isEnding()
  {
    return this.ending;
  }

  /**
   * Set a HTML attribute.
   * 
   * @param name
   *          The name of the attribute.
   * @param value
   *          The value of the attribute.
   */
  public void setAttribute(String name, String value)
  {
    this.attributes.put(name.toLowerCase(), value);
  }

  /**
   * @param ending
   *          The ending to set.
   */
  public void setEnding(boolean ending)
  {
    this.ending = ending;
  }

  /**
   * Set the tag name.
   */
  public void setName(String s)
  {
    this.name = s;
  }

  /**
   * Convert this tag back into string form, with the
   * beginning < and ending >.
   * 
   * @param id
   *          A zero based index that specifies the
   *          attribute to retrieve.
   * @return The Attribute object that was found.
   */
  @Override
  public String toString()
  {
    StringBuilder buffer = new StringBuilder("<");
    buffer.append(this.name);

    Set<String> set = this.attributes.keySet();
    for (String key : set)
    {
      String value = this.attributes.get(key);
      buffer.append(' ');

      if (value == null)
      {
        buffer.append("\"");
        buffer.append(key);
        buffer.append("\"");
      } else
      {
        buffer.append(key);
        buffer.append("=\"");
        buffer.append(value);
        buffer.append("\"");
      }

    }

    if (this.ending)
    {
      buffer.append('/');
    }
    buffer.append(">");
    return buffer.toString();
  }
}
